src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandWrapperJsonConverter.cs

```diff
@@ -0,0 +1,308 @@
+                    typeof(SelectExpandWrapperOfTJsonConverter<>).MakeGenericType(new Type[] { type }));
+            }
+
+            return null;
+        }
+
+        private class JsonPropertyNameMapper : IPropertyMapper
+        {
+            private IEdmModel _model;
+            private IEdmStructuredType _type;
+
+            public JsonPropertyNameMapper(IEdmModel model, IEdmStructuredType type)
+            {
+                _model = model;
+                _type = type;
+            }
+
+            public string MapProperty(string propertyName)
+            {
+                IEdmProperty property = _type.Properties().Single(s => s.Name == propertyName);
+                PropertyInfo propertyInfo = GetPropertyInfo(property);
+
+                Newtonsoft.Json.JsonPropertyAttribute jsonPropertyAttribute;
+                JsonPropertyNameAttribute jsonPropertyNameAttribute;
+
+                // NOTE: Preceding .NET Core 3.1 Newtonsoft was the default serialization library
+                // We check the Newtonsoft JsonPropertyAttribute first to prevent breaking changes
+                if ((jsonPropertyAttribute = GetJsonPropertyAttribute(propertyInfo)) != null && !string.IsNullOrWhiteSpace(jsonPropertyAttribute.PropertyName))
+                {
+                    return jsonPropertyAttribute.PropertyName;
+                }
+                else if ((jsonPropertyNameAttribute = GetJsonPropertyNameAttribute(propertyInfo)) != null && !string.IsNullOrWhiteSpace(jsonPropertyNameAttribute.Name))
+                {
+                    return jsonPropertyNameAttribute.Name;
+                }
+                else
+                {
+                    return property.Name;
+                }
+            }
+
+            private PropertyInfo GetPropertyInfo(IEdmProperty property)
+            {
+                ClrPropertyInfoAnnotation clrPropertyAnnotation = _model.GetAnnotationValue<ClrPropertyInfoAnnotation>(property);
+                if (clrPropertyAnnotation != null)
+                {
+                    return clrPropertyAnnotation.ClrPropertyInfo;
+                }
+
+                ClrTypeAnnotation clrTypeAnnotation = _model.GetAnnotationValue<ClrTypeAnnotation>(property.DeclaringType);
+                Contract.Assert(clrTypeAnnotation != null);
+
+                PropertyInfo info = clrTypeAnnotation.ClrType.GetProperty(property.Name);
+                Contract.Assert(info != null);
+
+                return info;
+            }
+
+            private static Newtonsoft.Json.JsonPropertyAttribute GetJsonPropertyAttribute(PropertyInfo property)
+            {
+                return property.GetCustomAttributes(typeof(Newtonsoft.Json.JsonPropertyAttribute), inherit: false)
+                    .OfType<Newtonsoft.Json.JsonPropertyAttribute>().SingleOrDefault();
+            }
+
+            private static JsonPropertyNameAttribute GetJsonPropertyNameAttribute(PropertyInfo property)
```

Comment:
Feel free to ignore this, but I think it would be less confusing for the names if something about the library is mentioned in the name, e.g. `GetNewtonsoftJsonPropertyAttribute`, Get `GetSystemJsonPropertyNameAttribute`